### PR TITLE
9981 builder redirections

### LINKED ISSUE #9981: Builder and dataset redirections
So:

- If a user has viz.json v2 visualizations, they should be opened with the Editor. Public maps should be opened with the public editor map, and builder map can't be reached.
- If a user has "migrated" its map to Builder, public or embed Editor map should redirect to Builder public map.

What about the table/dataset view?
- If a user has Builder enabled, dataset view.
- If a user has Editor enabled, table view (GME users for example).
- For public viewers, table view

Take into account this [premise](https://github.com/CartoDB/cartodb/issues/9978).

### PR DESCRIPTION
**Pull request by javitonino**
- Embed redirections (based on visualization version flag)
 - Auto migration (version update) for safe visualizations
 - Open vizjson2 visualizations in editor

Closes #9981

### CODE REVIEW

**Comment by javitonino** [comment]
Please CR @gfiorav

File: app/controllers/admin/visualizations_controller.rb [hunk 1]
```diff
@@ -36,6 +36,9 @@ class Admin::VisualizationsController < Admin::AdminController
                                                          :show_protected_public_map, :show_protected_embed_map]
 
   before_filter :resolve_visualization_and_table_if_not_cached, only: [:embed_map]
+  before_filter :redirect_to_builder_embed_if_v3, only: [:embed_map, :show_organization_public_map,
```
Comment (gfiorav) [inline app/controllers/admin/visualizations_controller.rb]: Test: 

0. Password protected map on a non migrated viz
1. Go to password prompt (for v2)
2. Put wrong password

Are you properly redirected?

File: app/controllers/admin/visualizations_controller.rb [hunk 2]
```diff
@@ -693,4 +696,12 @@ def track_dashboard_visit
                                                     user_id: current_user_id,
                                                     page: 'dashboard').report
   end
+
+  def redirect_to_builder_embed_if_v3
+    # @visualization is not loaded if the embed is cached
```
Comment (gfiorav) [inline app/controllers/admin/visualizations_controller.rb]: So if cached there will be no auto migration?

File: app/controllers/carto/admin/visualization_public_map_adapter.rb [hunk 3]
```diff
@@ -20,7 +20,8 @@ class VisualizationPublicMapAdapter
         :password_protected?, :varnish_key, :related_tables, :password_valid?, :get_auth_tokens, :table, :name,
         :overlays, :created_at, :updated_at, :description, :mapviews, :geometry_types, :privacy, :tags,
         :surrogate_key, :has_password?, :total_mapviews, :is_viewable_by_user?, :is_accesible_by_user?,
-        :can_be_cached?, :is_privacy_private?, :source, :kind_raster?, :has_read_permission?, :has_write_permission?
+        :can_be_cached?, :is_privacy_private?, :source, :kind_raster?, :has_read_permission?, :has_write_permission?,
+        :open_in_editor?, :version
```
Comment (gfiorav) [inline app/controllers/carto/admin/visualization_public_map_adapter.rb]: Minor, off topic: we should really stop worsening our viz<->map identity problem with all these delegates 😂

File: app/controllers/carto/builder/public/embeds_controller.rb [hunk 4]
```diff
@@ -8,7 +8,8 @@ class EmbedsController < BuilderController
 
         ssl_required :show, :show_protected
 
-        before_filter :load_visualization, only: [:show, :show_protected]
+        before_filter :load_visualization,
+                      :redirect_to_old_embed_if_v2, only: [:show, :show_protected]
```
Comment (gfiorav) [inline app/controllers/carto/builder/public/embeds_controller.rb]: Same testing proposed (wrong password on embed protected when not yet migrated).

File: app/controllers/carto/builder/visualizations_controller.rb [hunk 5]
```diff
@@ -12,8 +12,9 @@ class VisualizationsController < BuilderController
 
       ssl_required :show
 
-      before_filter :redirect_to_editor_if_forced,
-                    :load_derived_visualization, only: :show
+      before_filter :load_derived_visualization,
```
Comment (gfiorav) [inline app/controllers/carto/builder/visualizations_controller.rb]: why the change of order? curious

File: app/models/carto/overlay.rb [hunk 6]
```diff
@@ -23,6 +23,9 @@ class Overlay < ActiveRecord::Base
       'header', 'search', 'layer_selector', 'share', 'zoom', 'logo', 'loader', 'fullscreen'
     ].freeze
 
+    BUILDER_COMPATIBLE_TYPES = ['search', 'layer_selector', 'share', 'fullscreen', 'loader', 'logo', 'zoom'].freeze
+    scope :builder_incompatible, where('type NOT IN (?)', BUILDER_COMPATIBLE_TYPES)
```
Comment (gfiorav) [inline app/models/carto/overlay.rb]: nice trick!

File: app/models/visualization/member.rb [hunk 7]
```diff
@@ -364,6 +364,11 @@ def tags=(tags)
         super(tags)
       end
 
+      def version=(version)
+        self.dirty = true
```
Comment (gfiorav) [inline app/models/visualization/member.rb]: explain!

File: spec/models/table_spec.rb [hunk 8]
```diff
@@ -736,9 +736,9 @@ def @data_import.data_source=(filepath)
 
     id = table.table_visualization.id
     CartoDB::Varnish.any_instance.expects(:purge)
-                                 .times(4)
-                                 .with(".*#{id}:vizjson")
-                                 .returns(true)
+                    .at_least_once
```
Comment (gfiorav) [inline spec/models/table_spec.rb]: This test is now vulnerable to overlooking missing invalidations!

File: spec/models/table_spec.rb [hunk 9]
```diff
@@ -2298,7 +2298,7 @@ def @data_import.data_source=(filepath)
       CartoDB::Visualization::Member.stubs(:new).with(has_entry(id: derived.id)).returns(derived)
       CartoDB::Visualization::Member.stubs(:new).with(has_entry(type: 'table')).returns(table.table_visualization)
 
-      derived.expects(:invalidate_cache).once
+      derived.expects(:invalidate_cache).at_least_once
```
Comment (gfiorav) [inline spec/models/table_spec.rb]: Overlooking posible redundant invalidations. Although super minor.

File: spec/requests/admin/visualizations_spec.rb [hunk 10]
```diff
@@ -161,6 +161,36 @@ def polymorphic_path(*args)
...
           last_response.status.should eq 200
         end
+
+        it 'never for vizjson2 visualizations' do
+          @user.stubs(:builder_enabled).returns(true)
+          @user.stubs(:builder_enabled?).returns(true)
+          Carto::Visualization::any_instance.stubs(:uses_vizjson2?).returns(true)
+
+          login_as(@user, scope: @user.username)
+          get "/viz/#{@id}", {}, @headers
```
Comment (gfiorav) [inline spec/requests/admin/visualizations_spec.rb]: No path names? Helpful in case we change routes in the future!

File: spec/requests/admin/visualizations_spec.rb [hunk 11]
```diff
@@ -161,6 +161,36 @@ def polymorphic_path(*args)
...
+        it 'embed redirects to builder for v3 when needed' do
+          # These two tests are in the same testcase to test proper embed cache invalidation
+          @user.stubs(:builder_enabled).returns(false)
+          @user.stubs(:builder_enabled?).returns(false)
+          visualization = CartoDB::Visualization::Member.new(id: @id).fetch
+          visualization.version = 2
+          visualization.store
+
+          login_as(@user, scope: @user.username)
+          get "/viz/#{@id}/embed_map", {}, @headers
```
Comment (gfiorav) [inline spec/requests/admin/visualizations_spec.rb]: Same comment about routes.

File: app/controllers/carto/builder/visualizations_controller.rb [hunk 12]
```diff
@@ -49,7 +50,9 @@ def builder_notifications
       end
 
       def redirect_to_editor_if_forced
-        redirect_to CartoDB.url(self, 'public_visualizations_show_map', id: params[:id]) if current_user.force_editor?
+        if current_user.force_editor? || @visualization.open_in_editor?
```
Comment (gfiorav) [inline app/controllers/carto/builder/visualizations_controller.rb]: I've checked the code for `.force_editor?`. Why isn't it the same as `!.force_builder?`

File: app/controllers/carto/admin/visualization_public_map_adapter.rb (hunk 3 above)
Comment (javitonino) [inline app/controllers/carto/admin/visualization_public_map_adapter.rb]: This is kind of another thing. This is a wrapper around Carto::Visualization to make it look like a Memberino, so it can be used in controllers/presenters without the need to rewrite everything. ¯\_(ツ)_/¯

File: app/controllers/admin/visualizations_controller.rb (hunk 2 above)
Comment (javitonino) [inline app/controllers/admin/visualizations_controller.rb]: Auto migration only happens from the builder, never from the embed. Summarizing:

- Embed redirections only check the version field. 2 -> old embed, 3 -> new embed
- Builder redirections check:
  - User is builder (if not -> editor)
  - Visualization uses vizjson 2 (if yes -> editor)
  - Can be automigrated (if yes -> do it!)
  - Else: show builder with version = 2 -> Opens manual migration modal

File: app/controllers/carto/builder/public/embeds_controller.rb (hunk 4 above)
Comment (javitonino) [inline app/controllers/carto/builder/public/embeds_controller.rb]: Working!

File: app/controllers/carto/builder/visualizations_controller.rb (hunk 5 above)
Comment (javitonino) [inline app/controllers/carto/builder/visualizations_controller.rb]: Before, the redirection only depended on the user. Now, it also depends on the visualization itself (it uses vizjson2 or not). So I need to load the visualization before deciding to redirect. I could have kept the order and added a new filter for the viz-based redirection, but I prefer having all the redirection logic in the same place.

File: app/controllers/carto/builder/visualizations_controller.rb (hunk 12 above)
Comment (javitonino) [inline app/controllers/carto/builder/visualizations_controller.rb]: Because there is the option where nothing is force (`User.builder_enabled` is nil) and you can go to either editor and builder depending on the URL (ooooold behaviour). This will be removed soon-ish (next week most likely).

File: spec/models/table_spec.rb (hunk 8 above)
Comment (javitonino) [inline spec/models/table_spec.rb]: Uhm, didn't think that through. You are right! Changing back to 5.

File: spec/models/table_spec.rb (hunk 9 above)
Comment (javitonino) [inline spec/models/table_spec.rb]: I think this one is alright, as it is checking that it's invalidated. It shouldn't be a problem if it gets invalidated twice.

File: spec/requests/admin/visualizations_spec.rb (hunk 10 above)
Comment (javitonino) [inline spec/requests/admin/visualizations_spec.rb]: I'm lazy and copied it from the other specs. Changing.

File: spec/requests/admin/visualizations_spec.rb (hunk 11 above)
Comment (javitonino) [inline spec/requests/admin/visualizations_spec.rb]: Same

File: app/models/visualization/member.rb (hunk 7 above)
Comment (gfiorav) [inline app/models/visualization/member.rb]: Hacky! Can we make this into a method with a friendly name? 😁

**Comment by gfiorav** [comment]
Apart from that `self.dirty` stuff,

📦 it for the 🆓 🌍
